core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -161,7 +219,21 @@ private JoinNode getSizeBasedJoin(JoinNode joinNode, Context context)
 
         if (isLeftSideSmall) {
             // left side is small enough, but must be partitioned
-            return joinNode.flipChildren().withDistributionType(PARTITIONED);
+            return flippedJoin.withDistributionType(PARTITIONED);
+        }
+
+        // Flip join sides if left side is smaller than the right by more than SIZE_DIFFERENCE_THRESHOLD times.
+        // We use 8x factor because the reduction in size from aggregation and filtering
+        // may not be fully accounted here.
+        // We use getApproximateOutputSizeInBytes instead of getSourceTablesSizeInBytes to account for the effects of
+        // the operators between join and table scan as much as possible when comparing the size of the join sides.
+        double leftSourceSize = getApproximateOutputSizeInBytes(joinNode.getLeft(), context);
```

Comment:
Continuation of https://github.com/trinodb/trino/pull/11584#discussion_r844941518

> If the replication is due to right side being a scalar, then that's a good choice and this case was likely already captured in the previous cases about size < 100MB.

Yet we reached execution here, so previous cases didn't capture it, so we should assume replicated join shouldn't be chosen in absence of stats

> If the replication is due to cross-join then it'll be the same even if the join is flipped.

Left join without criteria must be replicated, but right join (flipped) without criteria must be partitioned (it actually lands on single node currently).
Anyway, I wouldn't make this method rely on `mustReplicate/mustPartition` internals

Reply:
I've updated this to not choose REPLICATED join explicitly, however I still need the `! mustReplicate` check to avoid incorrectly choosing PARTITIONED for a cross join.